File: public/services/actions/scorecard.ts
import { http, Result } from "@fider/services/http"
import { ScorecardField } from "@fider/models"

export interface ScorecardSettings {
  isEnabled: boolean
  bandStrong: number
  bandGood: number
  bandRefine: number
  bandLow: number
  triggerStatusSlug: string
}

export const updateScorecardSettings = async (settings: ScorecardSettings): Promise<Result> => {
  return await http.post("/_api/admin/scorecard-settings", settings)
}

export interface CreateScorecardFieldRequest {
  key: string
  label: string
  groupKey: string
  type: "text" | "note" | "date" | "number" | "url" | "choice" | "score"
  choices?: unknown
  weight?: number
  question?: string
  sortOrder: number
}

export const createScorecardField = async (req: CreateScorecardFieldRequest): Promise<Result<ScorecardField>> => {
  return await http.post<ScorecardField>("/_api/admin/scorecard-fields", req)
}

export interface UpdateScorecardFieldRequest {
  label: string
  choices?: unknown
  weight?: number
  question?: string
  sortOrder: number
  isActive: boolean
}

export const updateScorecardField = async (id: number, req: UpdateScorecardFieldRequest): Promise<Result> => {
  return await http.put(`/_api/admin/scorecard-fields/${id}`, req)
}

export const deleteScorecardField = async (id: number): Promise<Result> => {
  return await http.delete(`/_api/admin/scorecard-fields/${id}`)
}

export interface ScorecardRecord {
  id: number
  postId?: number
  title: string
  values: Record<string, unknown>
  createdAt: string
  updatedAt: string
}

export interface CreateScorecardRequest {
  postId?: number
  title?: string
}

export const createScorecard = async (req: CreateScorecardRequest): Promise<Result<ScorecardRecord>> => {
  return await http.post<ScorecardRecord>("/_api/scorecards", req)
}

export interface UpdateScorecardRequest {
  title: string
  values: Record<string, unknown>
}

export const updateScorecard = async (id: number, req: UpdateScorecardRequest): Promise<Result> => {
  return await http.put(`/_api/scorecards/${id}`, req)
}

File: app/models/cmd/scorecard.go
package cmd

import (
	"encoding/json"

	"github.com/getfider/fider/app/models/entity"
)

// CreateScorecardField inserts a new admin-defined field on the tenant's
// scorecard schema. Score-type rows must land in the scoring group. Weight
// and Question are only meaningful when Type == "score".
type CreateScorecardField struct {
	Key       string
	Label     string
	GroupKey  string
	Type      string
	Choices   json.RawMessage
	Weight    *int
	Question  string
	SortOrder int

	Result *entity.ScorecardField
}

// UpdateScorecardField mutates an existing field. For is_system=true rows the
// caller must still supply Key/Type/GroupKey but the postgres layer ignores
// them (locked). Label, Weight, Question, Choices, SortOrder, IsActive are
// always writable.
type UpdateScorecardField struct {
	ID        int
	Label     string
	Choices   json.RawMessage
	Weight    *int
	Question  string
	SortOrder int
	IsActive  bool
}

// DeleteScorecardField removes an admin-added field. Refused for is_system rows.
// Card values that referenced the field key stay in the JSONB blob — harmless,
// just no longer rendered.
type DeleteScorecardField struct {
	ID int
}

// SeedTenantScorecardFields seeds the 8 locked scoring dimensions for a newly
// created tenant. Idempotent — ON CONFLICT (tenant_id, key) DO NOTHING.
type SeedTenantScorecardFields struct {
	TenantID int
}

// CreateScorecard inserts a new scorecard optionally linked to a post. When
// PostID is set and a card already exists for that post, the handler is
// idempotent — it returns the existing card instead of creating a duplicate.
type CreateScorecard struct {
	PostID *int
	Title  string

	Result *entity.Scorecard
}

// UpdateScorecardValues replaces the values JSON blob for a scorecard, and
// optionally updates the display title. Tenant-scoped.
type UpdateScorecardValues struct {
	ID     int
	Title  string
	Values []byte // raw JSON blob
}

// DeleteScorecard removes a scorecard. Tenant-scoped, admin-only.
type DeleteScorecard struct {
	ID int
}

// SetTenantScorecardSettings toggles the feature on/off, sets the four band
// thresholds, and picks which status_slug triggers auto-create in one shot.
// Empty TriggerStatusSlug = no auto-trigger. Called by /admin/scorecard-settings.
type SetTenantScorecardSettings struct {
	IsEnabled         bool
	BandStrong        int
	BandGood          int
	BandRefine        int
	BandLow           int
	BandStrongLabel   string
	BandGoodLabel     string
	BandRefineLabel   string
	BandLowLabel      string
	BandNoneLabel     string
	TriggerStatusSlug string
}
